agent/testflinger_agent/stop_condition_checkers.py

```diff
@@ -0,0 +1,51 @@
+# Copyright (C) 2024 Canonical
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>
+
+import time
+from typing import Optional
+
+from .client import TestflingerClient
+
+
+class JobCancelledChecker:
+    def __init__(self, client: TestflingerClient, job_id: str):
+        self.client = client
+        self.job_id = job_id
+
+    def __call__(self) -> Optional[str]:
+        if self.client.check_job_state(self.job_id) == "cancelled":
+            return "\nJob cancellation was requested, exiting.\n"
+        return None
+
+
+class GlobalTimeoutChecker:
+    def __init__(self, timeout: int):
+        self.timeout = timeout
+        self.start_time = time.time()
+
+    def __call__(self) -> Optional[str]:
+        if time.time() - self.start_time > self.timeout:
+            return f"\nERROR: Global timeout reached! ({self.timeout}s)\n"
+        return None
+
+
+class OutputTimeoutChecker:
+    def __init__(self, timeout: int):
+        self.timeout = timeout
+        self.start_time = time.time()
+
+    def __call__(self) -> Optional[str]:
+        if time.time() - self.start_time > self.timeout:
+            return f"\nERROR: Output timeout reached! ({self.timeout}s)\n"
+        return None
```

Comment:
While start_time is initialized here, there are a few actions that happen before the process starts including some sleeping to wait for the process to start. Realistically I'm not sure how much difference there can be between start_time and the actual process start time. Perhaps it's not an issue. I was wondering however if there is some way to get the time spent from the process itself without measuring it ourselves.

Reply:
I'm not aware of a way to do that, but that would be very nice indeed! I don't think it's *too* big of a concern here. We're not working on the granularity of milliseconds here, but more like 15min+. If we spend even a few extra seconds before the process actually starts (unlikely) then I don't think that would be a problem.

Reply:
Makes sense, I found Popen.wait can take a timeout but that wouldn't work for OutputTimeoutChecker if I understand it correctly, so this solution suffices.